src/snowflake/snowpark/mock/pandas_util.py

```diff
@@ -0,0 +1,214 @@
+    quote_name_without_upper_casing,
+)
+from snowflake.snowpark._internal.type_utils import infer_type
+from snowflake.snowpark.exceptions import SnowparkClientException
+from snowflake.snowpark.table import Table
+from snowflake.snowpark.types import (
+    ArrayType,
+    BooleanType,
+    DecimalType,
+    DoubleType,
+    LongType,
+    MapType,
+    NullType,
+    StringType,
+    StructField,
+    StructType,
+    TimestampType,
+    VariantType,
+)
+
+if TYPE_CHECKING:
+    from snowflake.snowpark import DataFrame, Session
+
+
+PANDAS_INTEGER_TYPES = (
+    pd.Int8Dtype,
+    pd.Int16Dtype,
+    pd.Int32Dtype,
+    pd.Int64Dtype,
+    pd.UInt8Dtype,
+    pd.UInt16Dtype,
+    pd.UInt32Dtype,
+    pd.UInt64Dtype,
+)
+
+
+def _extract_schema_and_data_from_pandas_df(
+    data: pd.DataFrame,
+) -> Tuple[StructType, List[List[Any]]]:
+    """
+    infer column types from the pandas data
+    when running against snowflake, infer_schema (https://docs.snowflake.com/en/sql-reference/functions/infer_schema)
+    is used to infer schema.
+
+    pandas type related doc: https://pandas.pydata.org/docs/user_guide/basics.html#dtypes
+    """
+    col_names = [
+        quote_name_without_upper_casing(name) for name in data.columns.values.tolist()
+    ]
+    plain_data = [data.iloc[i].tolist() for i in range(data.shape[0])]
+    inferred_type_dict = (
+        {}
+    )  # this map is to store types for columns in which data are of primitive python objects
+    for row_idx in range(data.shape[0]):
+        for col_idx in range(data.shape[1]):
+            if plain_data[row_idx][col_idx] is None:
+                continue
+            if isinstance(plain_data[row_idx][col_idx], (float, numpy.float_)):
+                # in pandas, a float is represented in type numpy.float64
+                # which can not be inferred by snowpark python, we cast to built-in float type
+                # pandas PANDAS_INTEGER_TYPES (e.g. INT8Dtye) will also stored data in the format of float64
+                if math.isnan(plain_data[row_idx][col_idx]):
+                    plain_data[row_idx][col_idx] = None
+                else:
+                    plain_data[row_idx][col_idx] = (
+                        int(plain_data[row_idx][col_idx])
+                        if isinstance(data.dtypes[col_idx], PANDAS_INTEGER_TYPES)
+                        else float(plain_data[row_idx][col_idx])
+                    )
+            elif isinstance(plain_data[row_idx][col_idx], numpy.float32):
+                # convert str first and then to float to avoid precision drift as its stored in float32 format
+                plain_data[row_idx][col_idx] = float(str(plain_data[row_idx][col_idx]))
+            elif isinstance(plain_data[row_idx][col_idx], numpy.bool_):
+                plain_data[row_idx][col_idx] = bool(plain_data[row_idx][col_idx])
+            elif isinstance(plain_data[row_idx][col_idx], numpy.int_):
```

Comment:
What about the other primitive int types in https://numpy.org/doc/stable/user/basics.types.html#array-types-and-conversions-between-types?

Reply:
thanks for the doc, updated to use the more genric `numpy.unsignedinteger` and `numpy.signedinteger` for type checking here